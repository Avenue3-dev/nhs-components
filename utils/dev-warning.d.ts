/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export declare const useExperimentalWarning: (component: 'tabs' | 'character-count') => void;
/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export declare const useDevWarning: (message: string) => void;
