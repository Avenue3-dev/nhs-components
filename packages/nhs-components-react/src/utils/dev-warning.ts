const prod = process.env.NODE_ENV === 'production'

export const experimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  if (prod) {
    return
  }

  console.warn(`ATTENTION
  This component is tagged as experimental and therefore could change without notice. See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}
