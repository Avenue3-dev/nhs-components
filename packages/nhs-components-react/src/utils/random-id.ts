export const generateRandomString = (length = 5): string => {
  const rand = Math.random() + 1

  return rand.toString(36).substring(2, length + 2)
}

export const generateRandomName = (prefix?: string): string => {
  const randString = generateRandomString()

  return prefix ? `${prefix}_${randString}` : randString
}

export const generateRandomId = generateRandomName
