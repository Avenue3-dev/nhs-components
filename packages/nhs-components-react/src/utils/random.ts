export const generateRandomString = (length: number = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
