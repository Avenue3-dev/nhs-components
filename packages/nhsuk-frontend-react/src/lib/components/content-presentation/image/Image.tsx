import { HTMLProps } from 'react'
import clsx from 'clsx'

type Image = {
  Caption: typeof Caption
} & React.FC<ImageProps>

type ImageProps = {
  figureProps?: HTMLProps<HTMLDivElement>
} & HTMLProps<HTMLImageElement>

const Caption: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}): JSX.Element => {
  return (
    <figcaption className={clsx('nhsuk-image__caption', className)} {...rest}>
      {children}
    </figcaption>
  )
}

/**
 * Use images only if there is a real user need. Avoid unnecessary decoration.
 *
 * For more information on when to use this component, go to the [NHS Service Manuals website](https://service-manual.nhs.uk/design-system/components/images).
 *
 * @link https://service-manual.nhs.uk/design-system/components/images
 *
 * @example
 * ```tsx
 * <Image
 *  src="image.jpg"
 *  alt="Lots of sore red patches with small blisters spread across white skin on a woman&#39;s chest."
 * >
 * <Image.Caption>
 *  It can affect large areas of the body or limbs.
 * </Image.Caption>
 * </Image>
 * ```
 */
const Image: Image = ({ figureProps, children, className, alt, ...rest }) => {
  const { className: figureClassName, ...restFigureProps } = figureProps || {}

  return (
    <figure
      className={clsx('nhsuk-image', figureClassName)}
      {...restFigureProps}
    >
      <img
        className={clsx('nhsuk-image__img', className)}
        alt={alt}
        {...rest}
      />
      {children}
    </figure>
  )
}

Image.displayName = 'Image'
Caption.displayName = 'Image.Caption'

Image.Caption = Caption

export default Image
