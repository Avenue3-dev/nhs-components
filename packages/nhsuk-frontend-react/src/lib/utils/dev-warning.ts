const prod = process.env.NODE_ENV === 'production'

/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export const useExperimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  useDevWarning(`This component is tagged as experimental and therefore could change without notice.

  See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}

/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export const useDevWarning = (message: string): void => {
  if (prod) {
    return
  }

  console.warn(message)
}
