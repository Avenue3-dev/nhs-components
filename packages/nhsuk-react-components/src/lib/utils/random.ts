export const generateRandomString = (length = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
