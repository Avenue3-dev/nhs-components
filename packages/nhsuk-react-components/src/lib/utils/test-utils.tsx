/* eslint-disable import/export */
import {
  RenderOptions,
  RenderResult,
  cleanup,
  render,
} from '@testing-library/react'
import { MemoryRouter as Router } from 'react-router-dom'
import { afterEach } from 'vitest'

afterEach(() => {
  cleanup()
})

const customRender = (
  ui: React.ReactElement,
  options: RenderOptions = {},
): RenderResult => {
  return render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => <Router>{children}</Router>,
    ...options,
  })
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
