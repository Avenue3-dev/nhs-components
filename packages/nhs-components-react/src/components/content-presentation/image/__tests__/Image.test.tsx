import { render, screen } from '@/utils/test-utils'
import { describe, test } from 'vitest'
import Image from '..'

describe('Image', () => {
  test('should match snapshot', () => {
    render(
      <Image
        data-testid="image"
        src="image.jpg"
        alt="Lots of sore red patches with small blisters spread across white skin on a woman&#39;s chest."
      >
        <Image.Caption>
          It can affect large areas of the body or limbs.
        </Image.Caption>
      </Image>,
    )

    expect(screen.getByTestId('image').parentElement).toMatchSnapshot()
  })

  test('should match snapshot - no caption', () => {
    render(
      <Image
        data-testid="image"
        src="image.jpg"
        alt="Lots of sore red patches with small blisters spread across white skin on a woman&#39;s chest."
      />,
    )

    expect(screen.getByTestId('image').parentElement).toMatchSnapshot()
  })

  test('should match snapshot - figure props', () => {
    render(
      <Image
        data-testid="image"
        src="image.jpg"
        alt="Lots of sore red patches with small blisters spread across white skin on a woman&#39;s chest."
        figureProps={{
          className: 'custom-class',
        }}
      />,
    )

    expect(screen.getByTestId('image').parentElement).toMatchSnapshot()
  })
})
