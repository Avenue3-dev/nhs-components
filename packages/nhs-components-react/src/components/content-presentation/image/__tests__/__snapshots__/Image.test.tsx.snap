// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Image > should match snapshot - figure props 1`] = `
<figure
  class="nhsuk-image custom-class"
>
  <img
    alt="Lots of sore red patches with small blisters spread across white skin on a woman's chest."
    class="nhsuk-image__img"
    data-testid="image"
    src="image.jpg"
  />
</figure>
`;

exports[`Image > should match snapshot - no caption 1`] = `
<figure
  class="nhsuk-image"
>
  <img
    alt="Lots of sore red patches with small blisters spread across white skin on a woman's chest."
    class="nhsuk-image__img"
    data-testid="image"
    src="image.jpg"
  />
</figure>
`;

exports[`Image > should match snapshot 1`] = `
<figure
  class="nhsuk-image"
>
  <img
    alt="Lots of sore red patches with small blisters spread across white skin on a woman's chest."
    class="nhsuk-image__img"
    data-testid="image"
    src="image.jpg"
  />
  <figcaption
    class="nhsuk-image__caption"
  >
    It can affect large areas of the body or limbs.
  </figcaption>
</figure>
`;
